producer.go

```diff
@@ -394,48 +394,78 @@ func (p *Producer) leaderDispatcher(topic string, partition int32, input chan *M
 		output = p.getBrokerWorker(leader)
 	}
 
+	// highWatermark tracks the "current" retry level, which is the only one where we actually let messages through,
+	// all other messages get buffered in retryState[msg.retries].buf to preserve ordering
+	// retryState[msg.retries].expectChaser simply tracks whether we've seen a chaser message for a given level (and
+	// therefore whether our buffer is complete and safe to flush)
+	highWatermark := 0
```

Comment:
I suppose we could manage it by making `p.config.MaxRetries+1` the capacity instead of the length, and using `append` as appropriate. The annoying part is if we are no state at all (normal operation) and we get a message with `retries==9` or something then we have to loop adding empty elements to the array.

Not an unreasonable suggestion, but I think I like the current method better unless you have additional thoughts.